support-frontend/app/admin/settings/Settings.scala

```diff
@@ -35,27 +36,22 @@ object AllSettings {
 }
 
 object Settings {
-  private def fromBufferedSource[T: Decoder](buf: BufferedSource): Either[Throwable, T] = {
-    val settings = decode[T](buf.mkString)
-    Try(buf.close())
-    settings
-  }
 
-  def fromS3[T: Decoder](source: SettingsSource.S3)(implicit s3: AmazonS3): Either[Throwable, T] =
+  def fromS3[T: Decoder](source: SettingsSource.S3)(implicit s3: AwsS3Client): Either[Throwable, T] =
     for {
-      buf <- Either.catchNonFatal {
-        val inputStream = s3.getObject(source.bucket, source.key).getObjectContent
-        Source.fromInputStream(inputStream)
-      }.leftMap(ex => new RuntimeException(s"couldn't getObject content for source: $source", ex))
-      settings <- fromBufferedSource(buf)
+      buf <-
+        s3.fetchAsString(new AmazonS3URI("s3://" + source.bucket + "/" + source.key)).toEither
+      .leftMap(ex => new RuntimeException(s"couldn't getObject content for source: $source", ex))
+      settings <- decode[T](buf)
     } yield settings
 
   def fromLocalFile[T: Decoder](source: SettingsSource.LocalFile): Either[Throwable, T] =
     for {
       buf <- Either.catchNonFatal {
         Source.fromFile(source.path)
       }
-      settings <- fromBufferedSource(buf)
+      settings <- decode[T](buf.mkString)
+      _ <- Try(buf.close()).toEither
```

Comment:
I have changed all uses of S3 to use the new abstraction I created, which doesn't require you to close the stream yourself.  It is not actually `Future` compatible, but generally we seem to rely on blocking threads rather than async anyway. (which is a risk to server stability)

Reply:
It looks like in support-frontend this is only used for getting the settings, and this operation does have its own EC:
https://github.com/guardian/support-frontend/blob/master/support-frontend/app/admin/settings/SettingsProvider.scala#L134

An alternative would be to use `scala.concurrent.blocking`

Reply:
Is the blocking here necessary or just the way it's been written?

Reply:
not necessary, but InputStream is fundamentally a blocking abstraction and that's what AWS sdk exposes.
The only way around it if you want to stream data is to use some kind of reactive stream e.g. [fs2](https://github.com/functional-streams-for-scala/fs2) and an alternative AWS library e.g https://github.com/laserdisc-io/fs2-aws (from a quick google)
In our case, the files are small enough to fit in memory, so we would be happy with a Future[Array[Byte]] or similar, but under the covers this would still have to use some non blocking mechanism or a separate thread context.

Whether that's a priority depends whether it causes an incident!  But as we add more and more to support-frontend app it does add to the risk.  I think higher priority would be move promotions dynamo access over to the new non-blocking AWS SDK.